src/ray/util/concurrent_flat_map.h

```diff
@@ -0,0 +1,150 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <utility>
+
+#include "absl/container/flat_hash_map.h"
+#include "ray/util/mutex_protected.h"
+
+namespace ray {
+
+// A flat map that's protected by a single mutex with read/write locks.
+//
+// Use when expecting one of the following:
+// - mininal thread contention
+// - trivially copyable values
+// - majority reads
+// If these don't describe your use case, prefer boost::concurrent_flat_map.
+template <typename KeyType, typename ValueType>
+class ConcurrentFlatMap {
+ public:
+  template <typename... Args>
+  explicit ConcurrentFlatMap(Args &&...args) : map_(std::forward(args)...) {}
+
+  ConcurrentFlatMap() = default;
+
+  void Reserve(size_t size) {
+    auto write_lock = map_.LockForWrite();
+    write_lock.Get().reserve(size);
+  }
+
+  template <typename KeyLike>
+  std::optional<ValueType> Get(const KeyLike &key) const {
+    std::optional<ValueType> result;
+    auto read_lock = map_.LockForRead();
+    const auto &map = read_lock.Get();
+    const auto it = map.find(key);
+    if (it != map.end()) {
+      result.emplace(it->second);
+    }
+    return result;
+  }
+
+  // Returns true if key existed and visitor was called.
+  // Visitor is called under a write lock so should not be heavy, otherwise prefer Get
+  // followed by InsertOrAssign.
+  template <typename KeyLike>
+  bool WriteVisit(const KeyLike &key, const std::function<void(ValueType &)> &visitor) {
+    auto write_lock = map_.LockForWrite();
+    auto &map = write_lock.Get();
+    const auto it = map.find(key);
+    if (it != map.end()) {
+      visitor(it->second);
+      return true;
+    }
+    return false;
+  }
+
+  // Returns true if key existed and visitor was called.
+  // Calls function under read lock so should not be heavy, otherwise prefer Get to copy
+  // out.
+  template <typename KeyLike>
+  bool ReadVisit(const KeyLike &key,
+                 const std::function<void(const ValueType &)> &visitor) const {
+    auto read_lock = map_.LockForRead();
+    const auto &map = read_lock.Get();
+    auto it = map.find(key);
+    if (it != map.end()) {
+      visitor(it->second);
+      return true;
+    }
+    return false;
+  }
+
+  // Applies visitor to all values in the map.
+  // Calls under read lock so should not be heavy, otherwise prefer GetMapClone to copy.
+  void ReadVisitAll(
+      const std::function<void(const KeyType &, const ValueType &)> &visitor) const {
+    auto read_lock = map_.LockForRead();
+    const auto &map = read_lock.Get();
+    for (const auto &[key, value] : map) {
+      visitor(key, value);
+    }
+  }
+
+  template <typename KeyLike>
+  bool Contains(const KeyLike &key) const {
+    auto read_lock = map_.LockForRead();
+    return read_lock.Get().contains(key);
+  }
+
+  /// Will insert the key value pair if the key does not exist.
```

Comment:
Part of the comment in the file starts with `///`, others starts with `//`.
I'm ok with either, but please the same coding style.